kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/batch/source/Kafka10Reader.java

```diff
@@ -50,61 +50,62 @@
 
 
   /**
-   * Construct using the json representation of the kafka request
+   * Construct a reader based on the given {@link KafkaRequest}.
    */
-  public KafkaReader(KafkaRequest request) {
+  Kafka10Reader(KafkaRequest request) {
     kafkaRequest = request;
-    currentOffset = request.getOffset();
-    lastOffset = request.getLastOffset();
+    currentOffset = request.getStartOffset();
+    lastOffset = request.getEndOffset();
 
     // read data from queue
     Properties properties = new Properties();
     properties.putAll(request.getConf());
     consumer = new KafkaConsumer<>(properties, new ByteArrayDeserializer(), new ByteArrayDeserializer());
-    fetch();
   }
 
-  public boolean hasNext() throws IOException {
+  @Override
+  public boolean hasNext() {
     if (currentOffset >= lastOffset) {
       return false;
     }
     if (messageIter != null && messageIter.hasNext()) {
       return true;
-    } else {
-      return fetch();
     }
+    return fetch();
   }
 
   /**
-   * Fetches the next Kafka message and stuffs the results into the key and value.
+   * Fetches the next Kafka message. The message key will be set into the given {@link KafkaKey} object, and the message
+   * payload will be returned.
    */
-  public KafkaMessage getNext(KafkaKey kafkaKey) throws IOException {
-    if (hasNext()) {
-      ConsumerRecord<byte[], byte[]> consumerRecord = messageIter.next();
-
-      byte[] keyBytes = consumerRecord.key();
-      byte[] value = consumerRecord.value();
-      if (value == null) {
-        LOG.warn("Received message with null message.payload with topic {} and partition {}",
-                 kafkaKey.getTopic(), kafkaKey.getPartition());
-      }
-
-      ByteBuffer payload = value == null ? ByteBuffer.wrap(EMPTY_BYTE_ARRAY) : ByteBuffer.wrap(value);
-      ByteBuffer key = keyBytes == null ? ByteBuffer.wrap(EMPTY_BYTE_ARRAY) :  ByteBuffer.wrap(keyBytes);
-
-      kafkaKey.clear();
-      kafkaKey.set(kafkaRequest.getTopic(), kafkaRequest.getPartition(), currentOffset,
-                   consumerRecord.offset() + 1);
-      kafkaKey.setMessageSize(value == null ? -1 : value.length);
-      currentOffset = consumerRecord.offset() + 1; // increase offset
-      return new KafkaMessage(payload, key);
-    } else {
-      return null;
+  @Override
+  public KafkaMessage getNext(KafkaKey kafkaKey) {
+    if (!hasNext()) {
+      throw new NoSuchElementException("No message is available");
     }
+
+    ConsumerRecord<byte[], byte[]> consumerRecord = messageIter.next();
+
+    byte[] keyBytes = consumerRecord.key();
+    byte[] value = consumerRecord.value();
+    if (value == null) {
+      LOG.warn("Received message with null message.payload with topic {} and partition {}",
```

Comment:
was in existing code, but this is better as:

Received message with null payload from partition {} of topic {}.

This also seems like an easy way to flood the log with thousands of warnings. Do you know if the value is ever expected to be null?

Reply:
Probably not logging at all.

Reply:
Apparently it could be null based on the Kafka source.